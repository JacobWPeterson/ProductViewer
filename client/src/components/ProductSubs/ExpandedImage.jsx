import React from 'react';
import ReactDOM from 'react-dom';
import styled from 'styled-components';
import FontAwesome from 'react-fontawesome';

const Wrapper = styled.div`
  position: fixed;
  display: table;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  z-index: 5;
  max-width: 70vw;
  max-height: 95vh;
  overflow: hidden;
  box-shadow: 0 19px 38px rgba(0,0,0,0.30), 0 15px 12px rgba(0,0,0,0.22);
`;

const Expanded = styled.img`
  ${(props) => props.zoomed && 'transform: scale(2.5)'};
  // top: ${(props) => props.y};
  // left: ${(props) => props.x};
  background: url(${(props) => props.src}) no-repeat 0 0 fixed;
  display: block;
  max-height: 95vh;
  cursor: ${(props) => (props.zoomed ? 'zoom-out' : 'zoom-in')};
  background-position-x: ${(props) => props.x};
  background-position-y: ${(props) => props.y};
`;

const ModalLeftArrow = styled.div`
  color: #e0e0e0;
  font-size: 1.75rem;
  z-index: 6;
  position: absolute;
  top: 44vh;
  left: .5vw;
  &:hover {
    color: #80ccc4;
    transform: scale(1.1);
  };
`;

const ModalRightArrow = styled.div`
  color: #e0e0e0;
  font-size: 1.75rem;
  z-index: 6;
  position: absolute;
  right: .5vw;
  top: 44vh;
  &:hover {
    color: #80ccc4;
    transform: scale(1.1);
  };
`;

const IconHolder = styled.div`
  display: flex;
  position: absolute;
  left: 50%;
  transform: translate(-50%, -50%);
  bottom: 2vh;
  z-index: 6;
`;

const SelectedIcon = styled.div`
  background-color: #80ccc4;
  border-radius: 50%;
  width: .25rem;
  height: .25rem;
  z-index: 6;
  border-width: 0;
  border-style: none;
  margin: 0 .25vw;
  padding: 2px;
  border: 3px solid #80ccc4;
`;

const Icon = styled.div`
  border-radius: 50%;
  width: .25rem;
  height: .25rem;
  z-index: 6;
  border-width: 0;
  border-style: none;
  margin: 0 .25vw;
  padding: 2px;
  border: 3px solid #aeaeae;
  &:hover { border: 3px solid #80ccc4; };
`;

class ExpandedImage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      zoomed: false,
      x: 0,
      y: 0,
    };
    this.expanded = React.createRef();
    this.imageClickHandler = this.imageClickHandler.bind(this);
    this.arrowClickHandler = this.arrowClickHandler.bind(this);
    this.iconClickHandler = this.iconClickHandler.bind(this);
    this.trackMouse = this.trackMouse.bind(this);
  }

  trackMouse() {
    const node = this.expanded.current;

    node.addEventListener('mousemove', (e) => {
      // console.log('hello');
      this.setState({
        x: -e.offsetX,
        y: -e.offsetY,
      });
      // node.style.setProperty('--x', `${-e.offsetX}px`);
      // node.style.setProperty('--y', `${-e.offsetY}px`);
    });
  }

  imageClickHandler(event) {
    event.stopPropagation();
    this.setState((prevState) => ({ zoomed: !prevState.zoomed }), this.trackMouse);
  }

  arrowClickHandler(event) {
    event.stopPropagation();
    this.props.clickNavHandler(event);
  }

  iconClickHandler(event) {
    event.stopPropagation();
    this.props.clickedThumb(event.target.id);
  }

  render() {
    return ReactDOM.createPortal(
      <Wrapper>
        {this.props.index !== 0 && !this.state.zoomed && <ModalLeftArrow onMouseDown={this.arrowClickHandler}><FontAwesome id="-1" name="angle-left" size="2x" /></ModalLeftArrow> }
        <Expanded
          key={this.props.id}
          src={this.props.src}
          alt={this.props.alt}
          zoomed={this.state.zoomed}
          ref={this.expanded}
          x={this.state.x}
          y={this.state.y}
          onMouseDown={(event) => this.imageClickHandler(event)}
        />
        {this.props.index !== this.props.images.length - 1 && !this.state.zoomed && <ModalRightArrow onMouseDown={this.arrowClickHandler}><FontAwesome id="1" name="angle-right" size="2x" /></ModalRightArrow> }
        {!this.state.zoomed
        && (
          <IconHolder onMouseDown={(event) => event.stopPropagation()}>
            {this.props.images.map((image, index) => {
              const uniqueID = this.props.id * index;
              if (this.props.index === index) {
                return (
                  <SelectedIcon type="button" id={index} key={Number(uniqueID)} onMouseDown={this.iconClickHandler} />
                );
              }
              if (this.props.index !== index) {
                return (
                  <Icon type="button" id={index} key={Number(uniqueID)} onMouseDown={this.iconClickHandler} />
                );
              }
            })}
          </IconHolder>
        )}
      </Wrapper>,
      document.getElementById('modal-root') || document.createElement('div'), // for testing purposes,
    );
  }
}

export default ExpandedImage;
