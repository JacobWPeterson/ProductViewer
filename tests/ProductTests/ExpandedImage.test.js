/* eslint-disable no-undef */
/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { mount } from 'enzyme';
import styled from 'styled-components';

import ExpandedImage from '../../client/src/components/ProductSubs/ExpandedImage.jsx';
import ProductTestData from '../ProductTestData.js';

const props = {
  info: {
    source: ProductTestData.productStyles.results[0].photos[0].url,
  },
  id: ProductTestData.productStyles.results[0].style_id,
  images: ProductTestData.productStyles.results[0].photos,
};

describe('ExpandedImage', () => {
  const ExpandedImageComponent = mount(<ExpandedImage {...props} debug />);

  it('Should render correctly', () => {
    expect(ExpandedImageComponent).toMatchSnapshot();
  });
});
