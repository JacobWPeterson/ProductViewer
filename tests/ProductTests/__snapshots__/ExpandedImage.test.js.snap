// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpandedImage Should render correctly 1`] = `
<ExpandedImage
  debug={true}
  id={70550}
  images={
    Array [
      Object {
        "thumbnail_url": "https://images.unsplash.com/photo-1552902865-b72c031ac5ea?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
        "url": "https://images.unsplash.com/photo-1552902865-b72c031ac5ea?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80",
      },
      Object {
        "thumbnail_url": "https://images.unsplash.com/photo-1492447105260-2e947425b5cc?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
        "url": "https://images.unsplash.com/photo-1492447105260-2e947425b5cc?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1650&q=80",
      },
      Object {
        "thumbnail_url": "https://images.unsplash.com/photo-1548133464-29abc661eb5c?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
        "url": "https://images.unsplash.com/photo-1548133464-29abc661eb5c?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80",
      },
      Object {
        "thumbnail_url": "https://images.unsplash.com/photo-1500340520802-1687634cbe38?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
        "url": "https://images.unsplash.com/photo-1500340520802-1687634cbe38?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80",
      },
      Object {
        "thumbnail_url": "https://images.unsplash.com/photo-1559304022-afbf28f53c4d?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
        "url": "https://images.unsplash.com/photo-1559304022-afbf28f53c4d?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1656&q=80",
      },
      Object {
        "thumbnail_url": "https://images.unsplash.com/photo-1554921148-83d8ceda2095?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=80",
        "url": "https://images.unsplash.com/photo-1554921148-83d8ceda2095?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80",
      },
    ]
  }
  info={
    Object {
      "source": "https://images.unsplash.com/photo-1552902865-b72c031ac5ea?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=668&q=80",
    }
  }
>
  <Portal
    containerInfo={
      <div>
        <div
          class="ExpandedImage__Wrapper-sc-1g7vqin-0 iJrbzo"
        >
          <div
            class="ExpandedImage__ModalLeftArrow-sc-1g7vqin-2 ZdNkZ"
          >
            <span
              aria-hidden="true"
              class="fa fa-angle-left fa-2x"
              id="-1"
            />
          </div>
          <img
            class="ExpandedImage__Expanded-sc-1g7vqin-1 iLUdGR"
            x="0"
            y="0"
          />
          <div
            class="ExpandedImage__ModalRightArrow-sc-1g7vqin-3 kMeIWk"
          >
            <span
              aria-hidden="true"
              class="fa fa-angle-right fa-2x"
              id="1"
            />
          </div>
          <div
            class="ExpandedImage__IconHolder-sc-1g7vqin-4 iGQpBV"
          >
            <div
              class="ExpandedImage__Icon-sc-1g7vqin-6 hFSqpB"
              id="0"
              type="button"
            />
            <div
              class="ExpandedImage__Icon-sc-1g7vqin-6 hFSqpB"
              id="1"
              type="button"
            />
            <div
              class="ExpandedImage__Icon-sc-1g7vqin-6 hFSqpB"
              id="2"
              type="button"
            />
            <div
              class="ExpandedImage__Icon-sc-1g7vqin-6 hFSqpB"
              id="3"
              type="button"
            />
            <div
              class="ExpandedImage__Icon-sc-1g7vqin-6 hFSqpB"
              id="4"
              type="button"
            />
            <div
              class="ExpandedImage__Icon-sc-1g7vqin-6 hFSqpB"
              id="5"
              type="button"
            />
          </div>
        </div>
      </div>
    }
  >
    <ExpandedImage__Wrapper>
      <div
        className="ExpandedImage__Wrapper-sc-1g7vqin-0 iJrbzo"
      >
        <ExpandedImage__ModalLeftArrow
          onMouseDown={[Function]}
        >
          <div
            className="ExpandedImage__ModalLeftArrow-sc-1g7vqin-2 ZdNkZ"
            onMouseDown={[Function]}
          >
            <FontAwesome
              id="-1"
              name="angle-left"
              size="2x"
            >
              <span
                aria-hidden={true}
                className="fa fa-angle-left fa-2x"
                id="-1"
              />
            </FontAwesome>
          </div>
        </ExpandedImage__ModalLeftArrow>
        <ExpandedImage__Expanded
          key="70550"
          onMouseDown={[Function]}
          x={0}
          y={0}
          zoomed={false}
        >
          <img
            className="ExpandedImage__Expanded-sc-1g7vqin-1 iLUdGR"
            onMouseDown={[Function]}
            x={0}
            y={0}
          />
        </ExpandedImage__Expanded>
        <ExpandedImage__ModalRightArrow
          onMouseDown={[Function]}
        >
          <div
            className="ExpandedImage__ModalRightArrow-sc-1g7vqin-3 kMeIWk"
            onMouseDown={[Function]}
          >
            <FontAwesome
              id="1"
              name="angle-right"
              size="2x"
            >
              <span
                aria-hidden={true}
                className="fa fa-angle-right fa-2x"
                id="1"
              />
            </FontAwesome>
          </div>
        </ExpandedImage__ModalRightArrow>
        <ExpandedImage__IconHolder
          onMouseDown={[Function]}
        >
          <div
            className="ExpandedImage__IconHolder-sc-1g7vqin-4 iGQpBV"
            onMouseDown={[Function]}
          >
            <ExpandedImage__Icon
              id={0}
              key="0"
              onMouseDown={[Function]}
              type="button"
            >
              <div
                className="ExpandedImage__Icon-sc-1g7vqin-6 hFSqpB"
                id={0}
                onMouseDown={[Function]}
                type="button"
              />
            </ExpandedImage__Icon>
            <ExpandedImage__Icon
              id={1}
              key="70550"
              onMouseDown={[Function]}
              type="button"
            >
              <div
                className="ExpandedImage__Icon-sc-1g7vqin-6 hFSqpB"
                id={1}
                onMouseDown={[Function]}
                type="button"
              />
            </ExpandedImage__Icon>
            <ExpandedImage__Icon
              id={2}
              key="141100"
              onMouseDown={[Function]}
              type="button"
            >
              <div
                className="ExpandedImage__Icon-sc-1g7vqin-6 hFSqpB"
                id={2}
                onMouseDown={[Function]}
                type="button"
              />
            </ExpandedImage__Icon>
            <ExpandedImage__Icon
              id={3}
              key="211650"
              onMouseDown={[Function]}
              type="button"
            >
              <div
                className="ExpandedImage__Icon-sc-1g7vqin-6 hFSqpB"
                id={3}
                onMouseDown={[Function]}
                type="button"
              />
            </ExpandedImage__Icon>
            <ExpandedImage__Icon
              id={4}
              key="282200"
              onMouseDown={[Function]}
              type="button"
            >
              <div
                className="ExpandedImage__Icon-sc-1g7vqin-6 hFSqpB"
                id={4}
                onMouseDown={[Function]}
                type="button"
              />
            </ExpandedImage__Icon>
            <ExpandedImage__Icon
              id={5}
              key="352750"
              onMouseDown={[Function]}
              type="button"
            >
              <div
                className="ExpandedImage__Icon-sc-1g7vqin-6 hFSqpB"
                id={5}
                onMouseDown={[Function]}
                type="button"
              />
            </ExpandedImage__Icon>
          </div>
        </ExpandedImage__IconHolder>
      </div>
    </ExpandedImage__Wrapper>
  </Portal>
</ExpandedImage>
`;
